### PR TITLE
Restricted custom fields

### PR DESCRIPTION
**Pull request by romcheg**
Add an optional managing group property to custom fields. If set,
adding, changing or deleting values of those custom fields will
be restricted to mebmerb of a specified groups only.

Custom field form will not show such custom field values to
unauthorized users.

### CODE REVIEW

File: src/ralph/lib/custom_fields/admin.py [hunk 1]
```diff
@@ -135,3 +135,33 @@ def changeform_view(
         return super().changeform_view(
             request, object_id, form_url, extra_context
         )
+
+    def _create_formsets(self, request, obj, change):
+        """Helper function to generate formsets for add/change_view."""
```
Comment (MarekBleschke) [inline src/ralph/lib/custom_fields/admin.py]: Please add comment that this is 99% not altered internal django method + comment around lines that are altered and why.

**Review by MarekBleschke — CHANGES_REQUESTED** [review]

File: src/ralph/lib/custom_fields/admin.py (hunk 1 above)
Comment (romcheg) [reply src/ralph/lib/custom_fields/admin.py]: Done

**Review by Blejwi — DISMISSED** [review]

File: src/ralph/lib/custom_fields/api/viewsets.py [hunk 2]
```diff
@@ -42,6 +44,12 @@ def _get_related_model_info(self):
         }
         return info
 
+    def _user_can_manage_customfield(self, user, custom_field):
+        return (
+            custom_field.managing_group is None or
+            custom_field.managing_group in user.groups.all()
```
Comment (ar4s) [inline src/ralph/lib/custom_fields/api/viewsets.py]: Did you try ``user.groups.filter(pk=custom_field.managing_group.pk).exists()``?

File: src/ralph/lib/custom_fields/forms.py [hunk 3]
```diff
@@ -39,6 +40,21 @@ def save(self, *args, **kwargs):
 
 
 class CustomFieldValueFormSet(BaseGenericInlineFormSet):
+
+    def __init__(
+        self, data=None, files=None, instance=None, save_as_new=None,
+        prefix=None, queryset=None, request=None, **kwargs
+    ):
```
Comment (ar4s) [inline src/ralph/lib/custom_fields/forms.py]: You can write ``def __init__(self, request=None, *args, **kwargs):`` if you only need ``request``. Also ``super`` clause should simplify to ``super().__init__(request=request, queryset=queryset, *args, **kwargs)``

File: src/ralph/lib/custom_fields/forms.py [hunk 4]
```diff
@@ -39,6 +40,21 @@ def save(self, *args, **kwargs):
...
+        queryset = queryset.filter(
+            Q(custom_field__managing_group__isnull=True) |
+            Q(custom_field__managing_group__in=request.user.groups.all())
+        )
+
+        super().__init__(
+            data=data, files=files, instance=instance, save_as_new=save_as_new,
```
Comment (ar4s) [inline src/ralph/lib/custom_fields/forms.py]: Where is ``request``?

File: src/ralph/lib/custom_fields/admin.py [hunk 5]
```diff
@@ -135,3 +135,42 @@ def changeform_view(
         return super().changeform_view(
             request, object_id, form_url, extra_context
         )
+
+    def _create_formsets(self, request, obj, change):
```
Comment (ar4s) [inline src/ralph/lib/custom_fields/admin.py:139]: Where did you use this function?

**Review by MarekBleschke — DISMISSED** [review]

**Review by ar4s — CHANGES_REQUESTED** [review]

File: src/ralph/lib/custom_fields/admin.py (hunk 5 above)
Comment (romcheg) [reply src/ralph/lib/custom_fields/admin.py:139]: It it used internally by the ModelAdmin class

File: src/ralph/lib/custom_fields/api/viewsets.py (hunk 2 above)
Comment (romcheg) [reply src/ralph/lib/custom_fields/api/viewsets.py]: Looks more efficient. Thanks!

File: src/ralph/lib/custom_fields/forms.py (hunk 3 above)
Comment (romcheg) [reply src/ralph/lib/custom_fields/forms.py]: +1

File: src/ralph/lib/custom_fields/forms.py (hunk 4 above)
Comment (romcheg) [reply src/ralph/lib/custom_fields/forms.py]: It is not expected there.

**Review by ar4s — DISMISSED** [review]

**Review by Blejwi — DISMISSED** [review]

**Review by Blejwi — APPROVED** [review]

**Comment by romcheg** [comment]
The commits have been squashed. Since no other changes have been made, I'm going to merge this with a single approve.

**Pull request merged**